### PR TITLE
Refactor user dropdown menu

### PR DESCRIPTION
**Pull request by gee05053**
This PR resolves issue https://github.com/lablup/backend.ai-webui-ossca-2023/issues/21

스크린샷 2023-08-24 오후 1 22 55

### CODE REVIEW

File: react/src/components/UserProfileSettingModal.tsx [hunk 1]
```diff
@@ -36,7 +61,7 @@ const UserProfileSettingModal: React.FC = () => {
       suspense: false,
     },
   );
-  const totpActivated = userInfo?.user.totp_activated;
+  let totpActivated = userInfo?.user.totp_activated;
```
Comment (yomybaby) [inline react/src/components/UserProfileSettingModal.tsx]:
```suggestion
  const totpActivated = userInfo?.user.totp_activated;
```

File: react/src/components/UserDropdownMenu.tsx [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    },
+    {
+      suspense: false,
+    },
+  );
+  const userRole = userInfo?.user.role;
+
+  const getUsername = () => {
+    let name =
+      baiClient.full_name?.replace(/\s+/g, '').length > 0
```
Comment (yomybaby) [inline react/src/components/UserDropdownMenu.tsx]: Let's use `_.trim`.

File: resources/i18n/ko.json [hunk 3]
```diff
@@ -26,7 +26,7 @@
       "MyAccountInformation": "사용자 정보 변경",
       "MyAccount": "사용자 정보 변경",
       "FullName": "사용자 이름",
-      "FullNameInvalid": "사용자 이름은 64자 이하로 입력하십시오",
+      "FullNameInvalid": "사용자 이름은 0자 이상 64자 이하로 입력하십시오",
```
Comment (yomybaby) [inline resources/i18n/ko.json]:
```suggestion
      "FullNameInvalid": "사용자 이름은 64자 이하로 입력하십시오",
```

File: src/components/backend-ai-user-dropdown-menu.ts
Comment (yomybaby) [inline src/components/backend-ai-user-dropdown-menu.ts:1]: `UserDropdownMenu` was made to replace `backed-ai-user-dropdown-menu`. Let's make it work without `backed-ai-user-dropdown-menu` and remove it.

**Review by yomybaby — CHANGES_REQUESTED** [review]
image
There is no gap between project select and username.

**Comment by gee05053** [comment]
@yomybaby 
스크린샷 2023-08-29 오후 4 21 08
I modified code according to your code!

Please check this reply!
https://github.com/lablup/backend.ai-webui/pull/1882#discussion_r1308289913

**Comment by yomybaby** [comment]
The dropdown does not toggle by clicking(always open), even with `trigger={['click']}`. It seems like a bug in the Dropdown. 🤔 Hmm.. The example from Ant Design works properly. I'll check whether this is related to the container in the shadow DOM.

**Comment by gee05053** [comment]
@yomybaby 
I tried changed this line, toggle function is working well!!
스크린샷 2023-09-01 오전 9 50 57
스크린샷 2023-09-01 오전 9 51 15

**Comment by yomybaby** [comment]
@gee05053  Your commit [01f437c](https://github.com/lablup/backend.ai-webui/pull/1882/commits/01f437c76b1a4d129ab8305e625bca5abf5725be) does not fix that problem. I fixed the bug using a different approach. 😁

File: react/src/components/UserProfileSettingModal.tsx [hunk 4]
```diff
@@ -22,212 +48,140 @@ const UserProfileSettingModal: React.FC = () => {
...
-      {
-        onSuccess: () => {
+  const onSubmit = () => {
+    form.validateFields().then((values) => {
+      userMutations.updateFullName(values.full_name, {
+        onSuccess: (newFullName) => {
           messageApi.open({
             type: 'success',
-            content: t('webui.menu.PasswordUpdated'),
+            content: t('webui.menu.FullnameUpdated'),
```
Comment (agatha197) [inline react/src/components/UserProfileSettingModal.tsx]: It always comes out even if I don't change the fullname. It seems better to modify this message to appear only when the fullname is changed.

File: react/src/components/UserDropdownMenu.tsx [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+  UserOutlined,
+  MailOutlined,
+  SecurityScanOutlined,
+  ExclamationCircleOutlined,
+  LockOutlined,
+  HolderOutlined,
+  FileTextOutlined,
+  LogoutOutlined,
+} from '@ant-design/icons';
+import { useDebounce, useToggle } from 'ahooks';
```
Comment (agatha197) [inline react/src/components/UserDropdownMenu.tsx]:
```suggestion
import { useDebounce } from 'ahooks';
```

please remove unused import :)

File: react/src/components/UserProfileSettingModal.tsx [hunk 6]
```diff
@@ -22,212 +48,140 @@ const UserProfileSettingModal: React.FC = () => {
...
-      new_password2: string;
-    }) => {
-      return baiClient.update_password(
-        values.old_password,
-        values.new_password,
-        values.new_password2,
-      );
+  const mutationToRemoveTotp = useTanMutation({
+    mutationFn: (email: string) => {
+      return baiClient.remove_totp(email);
```
Comment (agatha197) [inline react/src/components/UserProfileSettingModal.tsx]:
```suggestion
    mutationFn: () => {
      return baiClient.remove_totp();
```

It is not possible for general users to remove totp using email.

File: react/src/hooks/backendai.tsx [hunk 7]
```diff
@@ -33,3 +35,137 @@ export const useResourceSlots = () => {
...
+          ...v,
+          full_name: input,
+        }));
+      }
+    };
+    document.addEventListener('current-user-info-changed', handler);
+    return () => {
+      document.removeEventListener('current-user-info-changed', handler);
+    };
+  }, []);
```
Comment (agatha197) [inline react/src/hooks/backendai.tsx]:
```suggestion
  }, [baiClient]);
```

File: react/src/hooks/backendai.tsx [hunk 8]
```diff
@@ -33,3 +35,137 @@ export const useResourceSlots = () => {
...
+
+  const mutationToUpdateUserFullName = useTanMutation({
```
Comment (agatha197) [inline react/src/hooks/backendai.tsx:92]: I cannot modify the username more than once

**Review by agatha197 — CHANGES_REQUESTED** [review]
By the way, the form layout of user setting modal of the credential page is horizontal, but the form layout of `UserProfileSettingModal` is vertical. We need to unify the style.
- `UserSettingModal`
image
- `UserProfileSettingModal`
image

File: react/src/hooks/backendai.tsx (hunk 8 above)
Comment (yomybaby) [reply react/src/hooks/backendai.tsx:92]: I removed the request condition for updateFullName in the hook. It's better to decide whether to skip the request or not where it's used.
Thank you for reporting.

File: react/src/components/UserProfileSettingModal.tsx [hunk 9]
```diff
@@ -277,37 +258,84 @@ const UserProfileSettingModal: React.FC = () => {
...
+                      form.setFieldValue('totp_activated', true);
+                      modal.confirm({
+                        title: t('totp.TurnOffTotp'),
+                        icon: <ExclamationCircleFilled />,
+                        content: t('totp.ConfirmTotpRemovalBody'),
+                        okText: t('button.Yes'),
+                        okType: 'danger',
+                        cancelText: t('button.No'),
+                        onOk() {
+                          mutationToRemoveTotp.mutate(userInfo.email || '', {
```
Comment (agatha197) [inline react/src/components/UserProfileSettingModal.tsx]: This occurs compiled error @yomybaby 
image

**Review by agatha197 — CHANGES_REQUESTED** [review]

File: react/src/components/UserProfileSettingModal.tsx (hunk 9 above)
Comment (yomybaby) [reply react/src/components/UserProfileSettingModal.tsx]: It's already resolved. Sorry for forgetting to push.

**Review by yomybaby — APPROVED** [review]

**Review by agatha197 — APPROVED** [review]
LGTM!

**Pull request merged**